bionic/deriver.py

```diff
@@ -460,46 +411,227 @@ def is_resolved(self):
+            self._wait_on_in_progress_entries()
+        return len(self._ready_entries) != 0
+
+    def _active_entry(self):
+        assert len(self._ready_entries) != 0
+        next_entry = self._ready_entries.pop()
+        assert next_entry.stage == EntryStage.READY
+
+        next_entry.stage = EntryStage.ACTIVE
+        return next_entry
+
+    def _wait_on_in_progress_entries(self):
+        "Waits on any in-progress entry to finish."
+        futures = [entry.future for entry in self._in_progress_entries.values()]
+        finished_futures, _ = wait(futures, return_when=FIRST_COMPLETED)
+        for finished_future in finished_futures:
+            task_key = finished_future.result()
+            entry = self._entries_by_task_key[task_key]
+            entry.state.sync_after_subprocess_completion()
+            self._mark_entry_completed(entry)
+
+    def _mark_entry_ready(self, ready_entry):
+        assert ready_entry.stage in (EntryStage.NEW, EntryStage.BLOCKED)
+
+        ready_entry.stage = EntryStage.READY
+        self._ready_entries.append(ready_entry)
+
+    def _mark_entry_blocked(self, blocked_entry):
+        assert blocked_entry.stage == EntryStage.ACTIVE
+
+        blocked_entry.stage = EntryStage.BLOCKED
+        blocking_tks = blocked_entry.state.incomplete_dep_task_keys()
+        blockage = EntryBlockage(blocked_entry, blocking_tks)
         for blocking_tk in blocking_tks:
             self._blockage_lists_by_blocking_tk[blocking_tk].append(blockage)
 
-    def get_unblocked_by(self, completed_state):
-        """
-        Records the fact that a task state is complete, and yields all task states
-        that are newly unblocked.
-        """
+        self._ready_blocking_entries(blocked_entry)
+
+    def _mark_entry_in_progress(self, in_progress_entry, future):
+        assert in_progress_entry.stage == EntryStage.ACTIVE
+        assert in_progress_entry.future is None
+        assert in_progress_entry not in self._in_progress_entries
+
+        in_progress_entry.stage = EntryStage.IN_PROGRESS
+        in_progress_entry.future = future
+        self._in_progress_entries[
+            in_progress_entry.state.task.keys[0]
+        ] = in_progress_entry
+
+    def _mark_entry_completed(self, completed_entry):
+        assert completed_entry.stage == EntryStage.ACTIVE or EntryStage.IN_PROGRESS
 
-        for completed_tk in completed_state.task.keys:
+        if completed_entry.stage == EntryStage.IN_PROGRESS:
+            completed_entry.future = None
+            del self._in_progress_entries[completed_entry.state.task.keys[0]]
+
+        completed_entry.stage = EntryStage.COMPLETED
+        self._unblock_entries(completed_entry)
+
+    def _ready_blocking_entries(self, blocked_entry):
+        incomplete_dep_states = blocked_entry.state.incomplete_dep_states()
+        for incomplete_dep_state in incomplete_dep_states:
+            incomplete_entry = self._get_or_create_entry_for_state(incomplete_dep_state)
```

Comment:
Are there any assertions we can/should make about this entry?

Reply:
This entry can be a new entry if first seen, or an already `READY` or `IN_PROGRESS` entry (if the entry blocks multiple entries and was added before). I guess it can't be active or completed since the task state wouldn't be incomplete but not sure if that requires an assertion. What do you think?

Reply:
Maybe after the `if` block we could `assert incomplete_entry in (EntryStage.READY, EntryStage.IN_PROGRESS)`? Then we can be certain that we'll end up making progress on it.